### PR TITLE
Fixes for LLVM 3.9 on Win64

### PR DESCRIPTION
**Pull request by vchuravy**
This build upon #19138. 

- COFF still doesn't work on 3.9, so we need to force ELF
- our build chain doesn't have support for c++11 threads, so we need to patch their usage out
- On debug builds LLVM started emitting calls to the MSVC crt so fcb408e is a temporary work around until we have a better solution.
- The rest are small fixes and quality of live improvements. 2e65e83 should probably be dropped, but it is needed for GCC 5.3

## Known issues
- Test-log: https://gist.github.com/vchuravy/e1276ed02c7cac750239f7c83d76e138
- `LLVM ERROR: Program used external function '__atomic_store' which could not be resolved!` -> `libatomic` is only supported by us under Linux
  There is no support as far as I can tell with the Mingw 4.9.2 toolchain, but they might have added it with Mingw 5.3 (see https://sourceforge.net/projects/mingw/files/MinGW/Base/gcc/Version5/gcc-5.3.0-2/libatomic-5.3.0-2-mingw32-dll-1.tar.xz/download)

cc: @tkelman, @vtjnash

### CODE REVIEW

**Comment by vchuravy** [comment]
I should also add that I won't have much time for the rest of this week, so if anybody want to take this up. Please go ahead.

File: test/Makefile [hunk 1]
```diff
@@ -10,7 +10,7 @@ default: all
 
 # set hard and soft limits for spawn fd-exhaustion test
 ifeq ($(OS),WINNT)
-ULIMIT_TEST=
+ULIMIT_TEST=true
```
Comment (tkelman) [inline test/Makefile]: This isn't cygwin specific, this just wasn't tested correctly. I have a PR that fixes it.

File: src/codegen.cpp [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #include "llvm-version.h"
 #include "platform.h"
 #include "options.h"
-#if defined(_OS_WINDOWS_) && JL_LLVM_VERSION < 30900
+#if defined(_OS_WINDOWS_) && JL_LLVM_VERSION < 40000
```
Comment (vchuravy) [inline src/codegen.cpp]: @vtjnash might know the answer for that.

My hypothesis is that we were hoping that COFF support might be fixed eventually.

File: test/Makefile (hunk 1 above)
Comment (vchuravy) [inline test/Makefile]: Cool I am happy to drop this then.

**Comment by tkelman** [comment]
how about we cherry pick the first 2 of your commits into https://github.com/JuliaLang/julia/pull/19138 and merge that now, then figure out what else we need to fix the test failure?

**Comment by vchuravy** [comment]
Absolutely, please go ahead.

**Pull request closed without merging**